apache/cloudstack
# KVM: Improvements on upload direct download certificates

## Description
There are some issues reported on the direct download feature for KVM:
-  The API method 'uploadTemplateDirectDownloadCertificate' always returns 'success:true' despite the ceritificate has been properly uploaded or not
- The certificate provided as parameter on 'uploadTemplateDirectDownloadCertificate' is not verified before being sent to hosts. This causes that hosts could not set up the certificates and fail the operation. 

To tackle these issues, several improvements for the KVM direct download feature have been introduced:
- Perform sanity check before sending certificate to hosts:
   - Verify the provided certificate is valid (right format) and not expired
- Report cases of failure when uploading certificates to hosts.
   - Return 'success:false' if the certificate cannot be uploaded to the hosts and log the issue properly
- Not issue-related improvements
   - Move the KVM direct download logic to KVM code
   - Add marvin tests

## Types of changes


## Screenshots (if appropriate):

## How Has This Been Tested?
2 KVM hosts
- Upload valid certificate, verify certificate is imported on both hosts -> OK
- Upload invalid certificate -> ERROR
- Upload existing alias certificate -> ERROR
- Remove certificates
- Try direct download
